src/canister_sig_util/src/lib.rs

```diff
@@ -125,6 +131,63 @@ pub fn extract_raw_canister_sig_pk_from_der(pk_der: &[u8]) -> Result<Vec<u8>, St
     Ok(pk_der[(bitstring_offset)..].to_vec())
 }
 
+pub fn hash_bytes(value: impl AsRef<[u8]>) -> Hash {
+    let mut hasher = Sha256::new();
+    hasher.update(value.as_ref());
+    hasher.finalize().into()
+}
+
+#[derive(Serialize)]
+struct CanisterSig {
+    certificate: ByteBuf,
+    tree: HashTree,
+}
+
+/// Returns canister signature for the specified `seed` and `message_hash` by retrieving
+/// the witness from `sig_map`.  If the canister uses
+/// [certified_data](https://internetcomputer.org/docs/current/references/ic-interface-spec/#system-api-certified-data)
+/// to manage certified assets, the caller should provide also the root hash of the assets subtree
+/// (cf. `asset_util`-crate).
+/// The returned value (if found) is a CBOR-serialised `CanisterSig`.
+pub fn get_signature_as_cbor(
+    sig_map: &SignatureMap,
+    seed: &[u8],
+    message_hash: Hash,
+    maybe_certified_assets_root_hash: Option<Hash>,
+) -> Result<Vec<u8>, String> {
+    let certificate = data_certificate()
+        .ok_or("data certificate is only available in query calls".to_string())?;
+    let witness = sig_map
+        .witness(hash_bytes(seed), message_hash)
+        .ok_or("missing witness".to_string())?;
+
+    let witness_hash = witness.digest();
+    let root_hash = sig_map.root_hash();
+    if witness_hash != root_hash {
```

Comment:
Does this check (still) make sense? We rehash the pruned witness tree every time, but if did not match the (unpruned) signature map root hash we would simply have a bug in our implementation. I wonder if this is really best addressed by wasting resources in production, or if we should rather have better tests (unless we're already sufficiently covered).
Honestly, I think we should remove it here.

Reply:
Filed https://dfinity.atlassian.net/browse/GIX-2178